core/primitives/src/subtrie.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Child trie related struct
+
+use parity_codec::{Encode, Decode};
+use rstd::prelude::*;
+use crate::storage::well_known_keys::CHILD_STORAGE_KEY_PREFIX;
+#[cfg(feature = "std")]
+pub use impl_serde::serialize as bytes;
+
+/// keyspace type.
+pub type KeySpace = Vec<u8>;
+
+
+/// info related to parent trie.
+/// Full key of child trie storage location
+/// and size of the prefix of this location.
+pub type ParentTrie = Vec<u8>;
+
+/// temp function to keyspace data above the db level
+pub fn keyspace_in_prefix(ks: &KeySpace, prefix: &[u8], dst: &mut[u8]) {
+	assert!(dst.len() == keyspace_prefixed_expected_len(ks, prefix));
+	dst[..ks.len()].copy_from_slice(&ks);
+	dst[ks.len()..].copy_from_slice(prefix);
+}
+
+/// len of targeted prefix with keyspace
+pub fn keyspace_prefixed_expected_len(ks: &KeySpace, prefix: &[u8]) -> usize {
+	ks.len() + prefix.len()
+}
+
+/// keyspace and prefix with allocation
+pub fn keyspace_as_prefix_alloc(ks: &KeySpace, prefix: &[u8]) -> Vec<u8> {
+	let mut res = rstd::vec![0;keyspace_prefixed_expected_len(ks, prefix)];
+	keyspace_in_prefix(ks, prefix, res.as_mut());
+	res
+}
+
+/// `SubTrieReadRef` used for non changing state query
+/// so it is safe to build
+#[derive(Clone)]
+pub struct SubTrieReadRef<'a> {
+	/// subtrie unique keyspace
+	pub keyspace: &'a KeySpace,
+	/// subtrie root hash
+	pub root: Option<&'a [u8]>,
```

Comment:
Can you mention in the rustdoc why this can be `None` and, ideally, why we need it?